### PR TITLE
Deterministic Pact SPV test

### PR DESCRIPTION
**Pull request by slpopejoy**

### CODE REVIEW

File: src/Chainweb/Cut/Test.hs [hunk 1]
```diff
@@ -119,6 +123,29 @@ testMine n t payloadHash i c =
...
+    -> GenBlockTime
+    -- ^ block time generation function
+    -> BlockPayloadHash
+    -> ChainId
+    -> Cut
+    -> IO (Either MineFailure (T2 BlockHeader Cut))
+testCut wdb n t payloadHash i c =
+  give wdb $
+    forM (createNewCut' (const True) n (t c i) payloadHash i c) $ \p@(T2 h _) ->
+        p <$ insertWebBlockHeaderDb h
```
[the remark this answers is not in the document]
Comment (slpopejoy) [reply src/Chainweb/Cut/Test.hs]: It just does adjacent header validation https://github.com/kadena-io/chainweb-node/blob/8f57fc770f675d70bfec04000b9ff4f20f1f70f0/src/Chainweb/WebBlockHeaderDB.hs#L183

File: src/Chainweb/Cut/Test.hs [hunk 2]
```diff
@@ -143,9 +170,27 @@ createNewCut
...
+    checkPOW h = checkTarget (_blockTarget h) $ _blockPow h
+
+-- | Create a new block. Only produces a new cut but doesn't insert it into the
+-- chain database. Optional POW validation.
+--
+createNewCut'
+    :: HasCallStack
+    => HasChainId cid
+    => (BlockHeader -> Bool)
+    -- ^ POW validator
```
[the remark this answers is not in the document]
Comment (slpopejoy) [reply src/Chainweb/Cut/Test.hs]: It's possible, but that's not the intention. The point here is to use mocking techniques to avoid violating separation-of-concerns in unit tests.

Using `ChainwebVersion` is less desirable, as it is overloaded with three separate concerns:
1. A protocol version. This in itself (protocol version compliance) is not of interest to many subsystems.
2. A graph. This is relevant to multi-chain-aware subsystems.
3. A hashing algorithm/DA/difficulty regime. This is relevant to mining- and DA-concerned subsystems.

Notice how all three of these properties are _irrelevant_ to some single-chain subsystem that is not involved in mining or DA or protocols. 

Here, by mocking out the nonce validation, a test is able to precisely specify whether #3 is of concern. `(const True)` (maybe with a suitably literate type synonym) is the most accurate way to say that mining difficulty is irrelevant; in this case it is strictly better as it makes a test less brittle if later on it needs to change something regarding #2 or #1. 

Finally, leaving it to `ChainwebVersion` means that the only true safe way to use these functions is via a `mine` function, _in case_ the protocol requires mining. This necessarily involves a clock, randomness etc, and is specifically what I wanted to avoid.

File: src/Chainweb/Cut/TestBlockDb.hs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import Chainweb.WebBlockHeaderDB
+
+import Data.CAS
+import Data.CAS.RocksDB
+
+data TestBlockDb = TestBlockDb
+  { _bdbWebBlockHeaderDb :: WebBlockHeaderDb
+  , _bdbPayloadDb :: PayloadDb RocksDbCas
+  , _bdbCut :: MVar Cut
+  }
```
[the remark this answers is not in the document]
Comment (slpopejoy) [reply src/Chainweb/Cut/TestBlockDb.hs]: That's very interesting. This was the "natural" result of finding the common dependencies of the SPV test, combined with the `MVar Cut` to simply use "current cut" as a shortcut for having a "latest block" concept. 

But now that you mention it, it does resemble `CutDB` (swap a `TVar` in anyway)! 

In general I was very pleased with how easy this refactoring was, which speaks extremely highly of the underlying design.

[the remark this answers is not in the document]
Comment (slpopejoy) [reply src/Chainweb/Cut/TestBlockDb.hs]: Also I would be happy to do your suggested refactoring. In general I am very glad to be getting to know these parts of the code.

File: test/Chainweb/Test/Pact/SPV.hs [hunk 4]
```diff
@@ -170,6 +160,27 @@ withAll vv f = foldl' (\soFar _ -> with soFar) f (chainIds vv) []
...
+getCutOutputs :: TestBlockDb -> IO CutOutputs
+getCutOutputs (TestBlockDb _ pdb cmv) = do
+  c <- readMVar cmv
+  cutToPayloadOutputs c pdb
+
+runCut :: TestBlockDb -> WebPactExecutionService -> IO CutOutputs
+runCut bdb pact = do
+  forM_ (chainIds v) $ \cid -> do
+    ph <- getParentTestBlockDb bdb cid
+    -- print (_blockHeight ph, cid)
```
[the remark this answers is not in the document]
Comment (slpopejoy) [reply test/Chainweb/Test/Pact/SPV.hs]: This is a useful debug output so I wanted to keep it. Should I instead make it a `Debug` log?

File: test/Chainweb/Test/Pact/SPV.hs [hunk 5]
```diff
@@ -170,6 +160,27 @@ withAll vv f = foldl' (\soFar _ -> with soFar) f (chainIds vv) []
...
+
+addPayloadBlock :: TestBlockDb -> ChainId -> PayloadWithOutputs -> IO ()
+addPayloadBlock db = addTestBlockDb db (Nonce 0) (offsetBlockTime second)
+
+getCutOutputs :: TestBlockDb -> IO CutOutputs
+getCutOutputs (TestBlockDb _ pdb cmv) = do
+  c <- readMVar cmv
+  cutToPayloadOutputs c pdb
+
+runCut :: TestBlockDb -> WebPactExecutionService -> IO CutOutputs
```
[the remark this answers is not in the document]
Comment (slpopejoy) [reply test/Chainweb/Test/Pact/SPV.hs]: I'll comment and move it to `Chainweb.Test.Pact.Utils` alongside `testWebPactExecutionService`.

Do you think the `CutOutputs` is also generally useful?

File: test/Chainweb/Test/Pact/SPV.hs [hunk 6]
```diff
@@ -180,94 +191,44 @@ roundtrip
...
-                         (chainToMPA txGen1) sqlenv1s
-            syncPact cutDb pact1
+    tg <- newMVar mempty
+    pact <- testWebPactExecutionService v
+              (return wdb) (return pdb)
+              (chainToMPA' tg) sqlenvs
 
-            c0 <- _cut cutDb
+    -- cut 0: empty run (not sure why this is needed but test fails without it)
+    void $ runCut bdb pact
```
[the remark this answers is not in the document]
Comment (slpopejoy) [reply test/Chainweb/Test/Pact/SPV.hs]: Not sure, I'll investigate.

File: test/Chainweb/Test/Pact/SPV.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (larskuhtz) [reply test/Chainweb/Test/Pact/SPV.hs]: Just ignore the suggestion. I was just not sure if the comment was there intentionally or just forgotten.

File: test/Chainweb/Test/Pact/SPV.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (larskuhtz) [reply test/Chainweb/Test/Pact/SPV.hs]: Not sure how useful the outputs are. However, throwing in a `void` on the use side isn't a big deal.

File: test/Chainweb/Test/Pact/SPV.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (larskuhtz) [reply test/Chainweb/Test/Pact/SPV.hs]: it may also be just a corner case related to starting with the genesis blocks. In production networks there probably never will be normal user txs at single digit block heights.

File: src/Chainweb/Cut/Test.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (larskuhtz) [reply src/Chainweb/Cut/Test.hs]: `insertBlockHeaderDb`, which is used to implement `insert` in the `TreeDb` instance of `BlockHeaderDb` calls this function for each inserted block:

```Haskell
-- | Validate properties of the block header, throwing an exception detailing
-- the failures if any.
--
-- This doesn't include checks for
--
-- * MissingPayload
-- * MissingParent
-- * IncorrectPayloadHash
--
validateBlockHeaderM
    :: MonadThrow m
    => Time Micros
        -- ^ The current clock time
    -> BlockHeader
        -- ^ parent block header. The genesis header is considered its own parent.
    -> BlockHeader
        -- ^ The block header to be checked
    -> m ()
```

This function runs all intrinsic and inductive block header validation, including `prop_block_pow `.

File: test/Chainweb/Test/Pact/SPV.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (slpopejoy) [reply test/Chainweb/Test/Pact/SPV.hs]: on second thought I'll just remove it

File: test/Chainweb/Test/Pact/SPV.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (slpopejoy) [reply test/Chainweb/Test/Pact/SPV.hs]: Hmm ... the problem with this function is it hardcodes the requirements of `addPayloadBlock` above ...

File: src/Chainweb/Cut/Test.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (larskuhtz) [reply src/Chainweb/Cut/Test.hs]: Short reply:

the test already depends on being used with a chainweb version that uses a trivial POW target, because `insert` does POW validation. Passing `const True` seems thus redundant.

Long reply:

The consistency rules for a chainweb are part of the chainweb version, which are implemented by the validation rules. Validation includes:

1. block header validation (intrinsic and indicutive)
2. cut validation (braiding structure of chainweb)
3. payload validation (construction of payload Merkle tree, tx uniqueness, tx timing validation, gas, etc.)
4. pact validation (computation of outputs via pact validation)

In the original design we used parameterized test chainweb versions to tweak consistency properties for testing. In most cases properties are statically bound to a particular version. Currently only the graph can be changed freely on some test versions. But @fosskers and I have been discussing to add free more parameters, like for instance the block rate.

In the current code we don't follow this original design everywhere, but we use chainweb versions to disable certain validation properties. For instance most test versions already use a trivial POW target, which is equivalent to passing `const True` to the POW validator.

[the remark this answers is not in the document]
Comment (larskuhtz) [reply src/Chainweb/Cut/Test.hs]: The idea behind using the chainweb version to relax validation constraints is to gather all configurations in a single place (`Chainweb.Version`) and avoid accidentally disabling or changing consistency properties. We try to make it impossible (or at least hard) to run code with chainweb version `Mainnet01` (or any other production version) with some consistency rules altered or disabled.

[the remark this answers is not in the document]
Comment (slpopejoy) [reply src/Chainweb/Cut/Test.hs]: Hmm, I got confused by the check in the code here, I didn't realize `insert` did it as well.

In that case the indirection is indeed unnecessary. On the larger topic above, I might want to allow tests to set the target to `0` (irrespective of what comes out of `newHeader` and it's merkle hash), but that's for a later debate. I'll switch to using `createNewCut` and drop this function.

File: test/Chainweb/Test/Pact/SPV.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (slpopejoy) [reply test/Chainweb/Test/Pact/SPV.hs]: I'd need to add a "payload adder" indirection basically with the signature of `addPayloadBlock`. Do you think that isn't too wonky for promotion?

[the remark this answers is not in the document]
Comment (slpopejoy) [reply test/Chainweb/Test/Pact/SPV.hs]: I'll try it out and see.

File: src/Chainweb/Cut/TestBlockDb.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (larskuhtz) [reply src/Chainweb/Cut/TestBlockDb.hs]: Just be aware that it might lead you to the less pleasant territories of the underlying design :-) Namely the code for validating remote cuts and fetching dependencies. But separating that from a local-only cut db might be useful.
